proposals/TP2-dynamically-generated-cacheable-xds-resources.md

```diff
@@ -0,0 +1,754 @@
+  <tr>
+    <td>
+<code>{constraints:[
+  {
+    constraints:[
+      {key:"env" value:"prod"},
+      {key:"version" value:"v1"}
+    ]
+    match_type: MATCH_TYPE_AND
+  }
+]}</code>
+    </td>
+    <td>
+      <ul>
+      <li>does include the route for <code>env=prod</code>
+      <li>does include the route for <code>version=v1</code>
+      </ul>
+    </td>
+  </tr>
+
+</table>
+
+## Rationale
+
+This section documents limitations and design alternatives that we
+considered.
+
+### Limitation on Enhancing Matching in the Future
+
+One limitation of this design is that, because all xDS transport protocol
+implementations (clients, servers, and caching proxies) need to implement
+this matching behavior, it will be very difficult to add new matching
+behavior in the future.  Doing so will probably require some sort of
+client capability.
+
+Because of this, reviewers of this design are encouraged to carefully
+scrutinize the proposed matching semantics to ensure that they meet our
+expected needs.
+
+### Using Context Parameters
+
+We considered extending the context parameter mechanism from [xRFC
+TP1](TP1-xds-transport-next.md) to support flexible matching semantics,
+rather that its current exact-match semantics.  However, that approach had
+some down-sides:
+- It would not have solved the virality problem described in the "Background"
+  section above.
+- It would have made the new xDS naming scheme a prerequisite for using
+  the dynamic resource selection mechanism.  (The mechanism described in
+  this doc is completely independent of the new xDS naming scheme; it can
+  be used with the legacy xDS naming scheme as well.)
+
+### Stricter Matching to Avoid Ambiguity
+
+We could avoid much of the matching ambiguity described above by saying that
+a set of constraints must specify all keys present in the subscription
+request in order to match.  However, this would mean that if the client
+starts subscribing with a new key before the corresponding constraint is
+added on the resources on the server, then it will fail to match the
+existing resources.  In other words, the process would be:
+
+1. Add a variant of all resources on the server side with a constraint
+   for `version=v1` (in addition to all existing constraints).
+2. Change clients to start sending the new key.
+3. When all clients are updated, remove the resource variants that do
+   *not* have the new key.
+
+This will effectively require adding new keys on the server side first,
+which seems like a large burden on users.  It also seems fairly tricky
+for most users to get the exactly correct set of dynamic parameters on
+each resource variant, and if they fail to do it right, they will break
+their existing configuration.
+
+Ultimately, although this approach is more semantically precise, it is
+also considered too rigid and difficult for users to work with.
+
+## Implementation
+
+TBD (Will probably be implemented in gRPC before Envoy)
+
+## Open issues (if applicable)
+
+N/A
```

Comment:
One thing I'm missing in this xRFC is discussion on caching data structures. Since they are contraint based, this is not in general a simple map. Can we provide some recommended algorithms or reference pseudo-code?

Reply:
I've added a section above describing implementation considerations that illustrates a naive implementation and suggests an approach for optimization.